### PR TITLE
Use blob storage for indexedDB (#42)

### LINKED ISSUE #42: Web: Use blob storage for indexedDB
Using base64 for binary data results in 33% size overhead. This is unacceptable for my use case...

Also, if we use blobs, a valid blob URI can be returned from `write_blob`. The current `README` example does not work for the web platform right now, since the source url is invalid.

### PR DESCRIPTION
**Pull request by Azarattum**
## Fixes #42

Now `capacitor-blob-writer` stores actuals blobs in IndexedDB on the web platform, instead of converting them to base64 strings. Also `write_blob` function returns a valid URL for the web environment (now the README example actually works in a browser).

All the tests in the demo app have passed unchanged (tested web only):
```
starting tests
wrote 10 bytes in 76ms
wrote 10 bytes in 18ms
wrote 10 bytes in 9ms
wrote 10 bytes in 22ms
wrote 10 bytes in 14ms
wrote 10 bytes in 15ms
wrote 10 bytes in 25ms
wrote 10 bytes in 32ms
wrote 10 bytes in 41ms
wrote 5242880 bytes in 19ms
tests passed!
```

Docs and benchmarks have been updated accordingly. The performance gains are quite impressive! It even managed to save a **512 MB** file!
| Size          | Filesystem       | BlobWriter          |
|---------------|------------------|---------------------|
| 1 kilobyte    | 4ms              | 9ms                 |
| 1 megabyte    | 180ms            | 16ms                |
| 8 megabytes   | 1.5s             | 43ms                |
| 32 megabytes  | 5.2s             | 141ms               |
| 64 megabytes  | 10.5s            | 0.2s                |
| 512 megabytes | Error[3]         | 1.1s                |

This implementation does not remove the previous one, but uses it as a fallback. You might consider removing `Filesystem.appendFile` version entirely since blobs in IDB are widely supported now ([in chrome since 2014](https://developer.chrome.com/blog/blob-support-for-Indexeddb-landed-on-chrome-dev/)). Anyway, this decision is up to you.

Also this might be considered a breaking change since `Filesystem.readFile` now returns a blob instead of a string. So, the package should be versioned appropriately.

### CODE REVIEW

**Comment by diachedelic** [comment]
What happens when you run, say,

    let img = document.createElement("img");
    Filesystem.getUri({path}).then(function (result) {
        img.src = Capacitor.convertFileSrc(result.uri);
    });

Does it work?

**Comment by Azarattum** [comment]
@diachedelic It will do exactly what it did before the PR. Meaning, it doesn't work. It returns the path from a fake file system, which browser does not understand. This is ultimately a `Capacitor.Filesystem`'s issue. 

It is possible to make it work with the new implementation. But we need to modify `convertFileSrc` or `getUri` implementation. Maybe something like that:
```js
import { convertFileSrc } from "capacitor-blob-writer";

let img = document.createElement("img");
Filesystem.getUri({path}).then(function (result) {
    img.src = convertFileSrc(result.uri);
});
```

To implement this we need a way to tell if we are dealing with a blob file without reading it. Because now it looks exactly like a normal file. I guess we could add an extra property or mess around with `filetype` prop. Which one is better depends on the `stat` implementation which I haven't look into yet.

To summarize, although the behavior is broken, this PR does not affect it.

**Comment by diachedelic** [comment]
Is it even possible to get a URL to a Blob in IndexedDB without using `URL.createObjectURL`?

**Comment by diachedelic** [comment]
That is to say, a URL which could be set as `img.src`?

**Comment by Azarattum** [comment]
> Is it even possible to get a URL to a Blob in IndexedDB without using `URL.createObjectURL`?

I don't think so. This is exactly what `URL.createObjectURL` was created for. You do not have any URLs in indexedDB. It's just a database.

**Comment by Azarattum** [comment]
What I think might work:
```js
function convertFileSrc({path}) {
const { type } = Filesystem.stat({path});
if (type == "blob") {
  const { data } = Filesystem.readFile({path});
  return URL.createObjectURL(data);
} else {
  return Filesystem. convertFileSrc(path);
}
```

Where we set `type = "blob"` in advance.

**Comment by diachedelic** [comment]
Thanks. I think it makes sense for the application developer to keep track of whether they are storing files as strings or Blobs. Blobs on the web platform can be accessed via `Filesystem.readFile` and then turned into a URL via `URL.createObjectURL`, in place of `Filesystem.getUri` and `Capacitor.convertFileSrc` on native platforms.

I am now just thinking about the breakage.

**Comment by Azarattum** [comment]
I think you should just semantically version it as `2.0.0` to avoid breaking changes upon update. Developers who want the new features and performance improvements will be able to opt in by manually updating.

**Comment by Azarattum** [comment]
> What I think might work:
> ```js
> function convertFileSrc({path}) {
>   const { type } = Filesystem.stat({path});
>   if (type == "blob") {
>     const { data } = Filesystem.readFile({path});
>     return URL.createObjectURL(data);
>   } else {
>     return Filesystem.convertFileSrc(path);
>   }
> }
> ```
> 
> Where we set `type = "blob"` in advance.

If you think this is a good idea, I might try to properly implement this as a QOL feature. So we won't put it off completely on the developers.

**Comment by diachedelic** [comment]
The plugin's interface is currently very simple, it is just a single function. I would prefer not to complicate it with an alternative `convertFileSrc`.

**Comment by Azarattum** [comment]
I don't see it a much of a complication since we will still have the function as a default export, meanwhile adding an option for a url converter for those who needs it. I may even be called something like `convertBlobSrc` to not confuse it with the original implementation.

```js
import writeBlob, { convertBlobSrc } from "capacitor-write-blob";
```
Since the second import is completely optional I don't see how it would complicate the interface.

**Comment by Azarattum** [comment]
> Thanks. I think it makes sense for the application developer to keep track of whether they are storing files as strings or Blobs. Blobs on the web platform can be accessed via `Filesystem.readFile` and then turned into a URL via `URL.createObjectURL`, in place of `Filesystem.getUri` and `Capacitor.convertFileSrc` on native platforms.
> 
> 
> 
> I am now just thinking about the breakage.

Actually I think this is a good idea to switch implementation depending on a platform. This actually brings even more incentive to put the utility function into the library. Personally, I as an application developer prefer all system-specific functionality to be hidden away.

That said, given this change we would really want to remove "appendFile" fallback to avoid any incompatibility. As I said before it shouldn't cause any issues.

**Comment by diachedelic** [comment]
The problem with `URL.createObjectURL` is that it [requires](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL#memory_management) a complementary `URL.revokeObjectURL` to avoid memory leaks. Memory leaks can [very quickly](https://github.com/ionic-team/capacitor/issues/2379) result in termination of the WebView on iOS, which is really hard to debug.

If we build `URL.createObjectURL` into the plugin, I don't see how we can also build in `URL.revokeObjectURL`. That's why I'd prefer to leave it up to the application developer. They will know when they don't need the URL anymore.

**Comment by diachedelic** [comment]
You are right about removing fallback mode for the web platform.

**Comment by Azarattum** [comment]
Yeah, I was also concerned about `URL.revokeObjectURL`... I guess the best we can do is to document it, so the developers know they should use it. I added a note for that in the readme, not sure if that's enough though.

I get your point why you want leave that up to developers. But what about the `URL.createObjectURL` returned from `write_blob` function? Do you think it's fine there?

**Comment by diachedelic** [comment]
I am thinking that `write_blob` should no longer produce a URI.

**Comment by Azarattum** [comment]
> I am thinking that `write_blob` should no longer produce a URI.

I think it makes sense, since a user already has an access to a blob they write. So they do `URL.createObjectURL` themselves.

**Comment by Azarattum** [comment]
So, @diachedelic, are you going to merge this or you still considering something?

File: README.md [hunk 1]
```diff
@@ -131,19 +131,21 @@ I have compared the performance & stability of `Filesystem.writeFile` with `writ
 - [1] Crashes the WKWebView, which immediately reloads the page
 - [2] `Failed to load resource: WebKit encountered an internal error`
 
-### Google Chrome (MacBook Pro 2012)
-The plugin falls back to `Filesystem.appendFile` in the browser, so these results should be approximately equal.
+### Google Chrome (Desktop FX-4350)
+The plugin puts your blob directly[1][2] into indexedDB storage if your browser supports that, otherwise it falls back to `Filesystem.appendFile`.
```
Comment (diachedelic) [inline README.md:135]: I thought we were not using the fallback for web?

File: README.md [hunk 2]
```diff
@@ -131,19 +131,21 @@ I have compared the performance & stability of `Filesystem.writeFile` with `writ
...
-- [1] `DOMException: The serialized keys and/or value are too large`
+| 1 kilobyte    | 4ms              | 9ms                 |
+| 1 megabyte    | 180ms            | 16ms                |
+| 8 megabytes   | 1.5s             | 43ms                |
+| 32 megabytes  | 5.2s             | 141ms               |
+| 64 megabytes  | 10.5s            | 0.2s                |
+| 512 megabytes | Error[3]         | 1.1s                |
+
+- [1] Data returned from `Filesystem.readFile` is your pure blob (not base64 encoded)
+- [2] The returned uri is created with `URL.createObjectURL` from your blob (e.g. `blob:http://localhost/189603f7-b6e7-40f2-92e4-466dc875633b`)
```
Comment (diachedelic) [inline README.md:147]: This is vital information, which should not be hidden away in the footnotes of the benchmarks section.

File: blob_writer.js [hunk 3]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
     });
 }
 
+function update_blob_content(directory, relative, blob) {
+	const indexedDB = window.indexedDB ||
+        window.mozIndexedDB ||
+        window.webkitIndexedDB ||
+        window.msIndexedDB;
```
Comment (diachedelic) [inline blob_writer.js:53]: Is it necessary to include these vendor prefixes? They seem to be relevant only to old versions of evergreen browsers (https://caniuse.com/?search=IndexedDB).

File: blob_writer.js [hunk 4]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
...
+
+	return new Promise((resolve, reject) => {
+		const connection = indexedDB.open("Disc");
```
Comment (diachedelic) [inline blob_writer.js:56]: Should we include a workaround for this catastrophic iOS 14.1 bug?
https://bugs.webkit.org/show_bug.cgi?id=226547

File: blob_writer.js [hunk 5]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
...
+		const fail = ({ target }) => reject(target);
```
Comment (diachedelic) [inline blob_writer.js:57]: Please use named functions.

File: blob_writer.js [hunk 6]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
...
+
+		connection.onerror = fail;
+		connection.onsuccess = ({ target }) => {
+			const db = target.result;
```
Comment (diachedelic) [inline blob_writer.js:61]: Rather than using `event.target` everywhere, just use the original variable (in this case, `connection`).

File: blob_writer.js [hunk 7]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
...
+			transaction.onerror = fail;
+
+			const store = transaction.objectStore("FileStorage");
+			const path = `/${directory}/${relative.replace(/^\//, "")}`;
+
+			const request = store.get(path);
+			request.onerror = fail;
+			request.onsuccess = ({ target }) => {
+				const request = store.put({
+					...target.result,
```
Comment (diachedelic) [inline blob_writer.js:73]: What properties are we expecting to keep here?

File: blob_writer.js [hunk 8]
```diff
@@ -46,6 +46,52 @@ function append_blob(directory, path, blob) {
...
+					const request = store.get(path);
+					request.onerror = fail;
+					request.onsuccess = (event) => {
+						const { target } = event;
+						if (!target) fail(event);
+						const { result } = target;
+						if (!result) fail(event);
+						const { content } = result;
+						if (!content) fail(event);
+						return resolve(URL.createObjectURL(content));
```
Comment (diachedelic) [inline blob_writer.js:88]: `write_file` is no longer producing a URI.

File: blob_writer.js [hunk 9]
```diff
@@ -62,11 +108,12 @@ function write_file_via_bridge({
         recursive,
         data: ""
     }).then(function ({uri}) {
+        return update_blob_content(directory, path, blob).catch(async () => {
```
Comment (diachedelic) [inline blob_writer.js:111]: We only truncate the file so that we can run `append_blob` multiple times. For the web platform, we can just overwrite the blob directly. I would leave `write_file_via_bridge` as is, and create a new function `write_file_via_indexeddb` to handle the new functionality. Use `Capacitor.platform` to determine the strategy, rather than depending on exceptions.

**Review by diachedelic — CHANGES_REQUESTED** [review]
There are things left to do.

# README.md
- Correct the introductory paragraph.
- In the example, show how to get hold of the file's URI on both native and web platforms. Warn the reader about memory leaks.
- Update the "How it works" section.
- In the "Fallback mode" section, mention that the web platform never falls back.

# blob_writer.js
- Change `write_file` to produce `undefined`, rather than the file's URI.
- Run the file thru the linter, available at https://james.diacono.com.au/jslint/.

File: blob_writer.js (hunk 7 above)
Comment (Azarattum) [reply blob_writer.js:73]: Here we want to keep all the metadata generated by `Filesystem.writeFile`. Such as `ctime`, `mtime`, `folder`, `path` and `type`.

File: blob_writer.js (hunk 9 above)
Comment (Azarattum) [reply blob_writer.js:111]: We should use `Capacitor.getPlatform()` instead of `Capacitor.platform` since the later is deprecated.

**Comment by Azarattum** [comment]
@diachedelic, I've implemented the requested changes.

**Comment by diachedelic** [comment]
Thanks. Nice doing business with you!

**Comment by diachedelic** [comment]
This feature is available in v1.1.0. I realised that we could make the behaviour opt-in, thereby avoiding breakage.
